ow_power_system/src/power_system_node.cpp

```diff
@@ -129,23 +79,66 @@ int main(int argc, char* argv[]) {
-    } //publishes last voltage value if sim time exceeds the length of the csv
-
-    //publish current SOC
-    SOC_pub.publish(soc_msg);
-    RUL_pub.publish(rul_msg);
-    t_step+=1;
-    
-    ros::spinOnce();
-    rate.sleep();
-   
-  }
 
+    // For each line of data in the example file, run a single prediction step.
+    for (const auto & line : power_data) {
+      // Get a new prediction
+      auto prediction = prognoser->step(line);
+      // Get the event for battery EoD. The first line of data is used to initialize the observer,
+      // so the first prediction won't have any events.
+      if (prediction.getEvents().size() == 0) {
+          continue;
+      }
+      auto eod_event = prediction.getEvents().front();
+      // The time of event is a `UData` structure, which represents a data
+      // point while maintaining uncertainty. For the MonteCarlo predictor
+      // used by this example, the uncertainty is captured by storing the
+      // result of each particle used in the prediction.
+      UData eod_time = eod_event.getTOE();
+      if (eod_time.uncertainty() != UType::Samples) {
+        ROS_ERROR_NAMED("power_system_node", "Unexpected uncertainty type for EoD prediction");
+      }
+      
+      // For this example, we will print the median EoD.
+      auto samples = eod_time.getVec();
+      std::sort(samples.begin(), samples.end());
+      double eod_median = samples.at(samples.size() / 2);
+      auto now =  MessageClock::now();
+      auto now_s = duration_cast<std::chrono::seconds>(now.time_since_epoch());
+      double rul_median = eod_median - now_s.count();
+      rul_msg.data = rul_median;
+
+      // State of Charge Code
+      //UData currentSOC = eod_event.getState()[0];
+      // For this example, we will print the median SOC
+      //auto samples2 = currentSOC.getVec();
+      //std::sort(samples2.begin(), samples2.end());
+      //double soc_median = samples2.at(samples2.size() / 2);
+      //soc_msg.data = soc_median;
+
+      // Temperature Code
+      //std::vector<UData> systemStates = eod_event.getSystemState()[0]; // Get the system (battery) state at t=0 (now)
+      // This is in format [stateId][sampleId].
+      // Here you will have to choose the sampleId the corresponds to your chosen percentile (e.g., Median) - and load it into a vector<double> where each element is a different state (in same order as above). 
+      // For Example:
+      //std::vector<double> medianSystemState(systemStates.size());
+      //for (int i = 0; i < medianSystemState.size(); i++) 
+        //medianSystemState[i] = systemStates[i][MEDIAN]; // Where MEDIAN corresponds to the sampleID of the median EOL prediction
+        
+      // Next, You will pass it into the model outputEqn to get the outputs:
+      //auto output = model.outputEqn(medianSystemState);
+      //double temperature = output[1];
+      
+      //publish current SOC & RUL
+      //SOC_pub.publish(soc_msg);
```

Comment:
is SOC no longer being published here at all? If so where is that being moved to?

Reply:
Sorry, you missed this in the tagup meeting yesterday: the SOC value is inaccessible based on the current framework of GSAP. I spoke with the main author of the software, Chris Teubert, who informed me that a GSAP PR had been created to change this that will hopefully be merged soon. Same goes for temperature. Mike suggested, however, that in the meantime, it might be useful to create a PR for the updated RUL calculation (which actually uses the prognostics library rather than just reading from a csv), and then create a separate PR once those two values are available in GSAP. You'll notice that the code for SOC and temperature is there, but has been commented out. For now, you might use the existing code base, which includes the place-holding (albeit very inaccurate) SOC values, but keep in mind that my next PR will resolve this and the range of the SOC values should be between 0 and 1 (as opposed to the current values that are around 4)